File: src/main/java/io/github/mosaicmc/mosaiccore/mixins/MinecraftServerMixin.java
package io.github.mosaicmc.mosaiccore.mixins;

import io.github.mosaicmc.mosaiccore.MainKt;
import io.github.mosaicmc.mosaiccore.plugin.PluginContainer;
import net.minecraft.server.MinecraftServer;
import org.slf4j.LoggerFactory;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@Mixin(MinecraftServer.class)
public class MinecraftServerMixin {
    @Inject(at = @At(value = "INVOKE", target = "Lnet/minecraft/server/MinecraftServer;setupServer()Z"), method = "runServer")
    private void pluginLoader(CallbackInfo info) {
        final var beforePlugins = MainKt.getBeforePlugins();
        for(final var plugin : beforePlugins) {
            final var entryPoint = plugin.getEntrypoint();

            entryPoint.beforeLoad();
        }

        final var plugins = MainKt.getPlugins();
        for(final var plugin : plugins) {
            final var entryPoint = plugin.getEntrypoint();
            final var modContainer = plugin.getProvider();
            final var server = (MinecraftServer) (Object) this;

            entryPoint.onLoad(new PluginContainer(
                    modContainer,
                    server,
                    modContainer.getMetadata().getName(),
                    LoggerFactory.getLogger(modContainer.getMetadata().getName())
            ));
        }
    }
}

File: src/main/kotlin/io/github/mosaicmc/mosaiccore/mixins/MinecraftServerMixin.kt
/*
 * Copyright (c) 2023. JustFoxx
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package io.github.mosaicmc.mosaiccore.mixins

import io.github.mosaicmc.mosaiccore.logger
import io.github.mosaicmc.mosaiccore.plugin.PluginContainer
import io.github.mosaicmc.mosaiccore.plugins
import net.minecraft.server.MinecraftServer
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo


@Mixin(MinecraftServer::class)
class MinecraftServerMixin {
    @Inject(
        at = [At(value = "INVOKE", target = "Lnet/minecraft/server/MinecraftServer;setupServer()Z")],
        method = ["runServer"]
    )
    private fun pluginLoader(info: CallbackInfo) {
        for (plugin in plugins) {
            val entryPoint = plugin.entrypoint
            val modContainer = plugin.provider
            val server = this as MinecraftServer
            val pluginContainer = PluginContainer(modContainer, server)

            entryPoint.onLoad(pluginContainer)

            logger.info("loaded ${pluginContainer.name}:${pluginContainer.metadata.version}")
        }
    }
}
